### ISSUE TITLE
Implicit property mapping based on naming overrides explicit

### ISSUE THREAD

**Issue by bstrandjev** [modelmapper/modelmapper#283]
After I create the issue I will add code to demonstrate the issue

**Comment by bstrandjev** [comment]
[This commit](https://github.com/bstrandjev/modelmapper/commit/75d6590773b6656076a9d4c71473c31a77b9f221) features a test demonstrating the issue. It also includes my proposal for logic to fix it.

However, with these changes ProviderTest2#shouldMapWithPropertyProviderViaPropertyMap fails.

I do not entirely understand how providers should work and thus can not figure out the correct logic to address this regression.

**Comment by bstrandjev** [comment]
The issue in the test is that the `unverifiedRequester` field in the `Requester` class, matches the by name and the name of the mapped property in this explicit mapping:

`map(source.getRequester()).setRequester(null);`

And thus results in all requester properties being null in the resulting object (in the test the `unverifiedRequester` is never initialized and thus remains `null`)

**Comment by chhsiao90** [comment]
Thanks for this report, your case is a bit complicated, I will take a look, thanks.

**Comment by chhsiao90** [comment]
Hi @bstrandjev - I had a PR that fixed this problem.

My idea is, we should not traversal the destination properties if there is already a TypeMap defined in our TypeMapStore.
But with this PR, you should change the ordering of adding the mappings to the ModelMapper as below.

```java
modelMapper.addMappings(new PropertyMap<Requester, PersonDetails>() {
    @Override
    protected void configure() {
        map(source.getPerson().getFirstName()).setFirstName(null);
        map(source.getPerson().getLastName()).setLastName(null);
    }
});
modelMapper.addMappings(new PropertyMap<Order, OrderView>() {
    @Override
    protected void configure() {
        map(source.getRequester()).setRequester(null);
    }
});
```

Is this PR can fix your problem? Thanks!

**Comment by bstrandjev** [comment]
I confirm the test that I wrote passes with your PR. However:
 
1. I want to spend some more time to understand the nature of your change
1. I do not like the fact that the order of adding the mappings actually matter. The defect I created was a result of a bug that took me DAYS to localize. I definitely doubt that anyone stumbling upon the same problem will know from the very beginning that there is a particular order of adding the mappings that will work for him.

**Comment by chhsiao90** [comment]
Yeah, agreed with you. The ordering about TypeMap to be added into the modelmapper is also been discussed at #283.

I know it's not very straightforward, but it's just how ModelMapper works.
If TypeMap A depends on TypeMap B, then you should add TypeMap B before TypeMap A.

For your case, maybe you can try disabled the [setImplicitMappingEnabled](http://modelmapper.org/javadoc/org/modelmapper/config/Configuration.html#setImplicitMappingEnabled-boolean-) then you can ignore the ordering.

**Comment by bstrandjev** [comment]
Maybe you cited the wrong issue in 

>  is also been discussed at #283.

Can you please check?

The suggestion 

> For your case, maybe you can try disabled the setImplicitMappingEnabled then you can ignore the ordering.

Is not good either, I think, because it is yet another solution that you will identify only if you know what the issue is and how to fix it (which will most certainly not be the case for any person that hits the same problem as me).

I truly believe that the direction this should be solved with is to completely disallow implicit mappings overriding explicit ones. Such logic should not depend on the ordering of adding mappings, as the implicit mappings are only calculated after all explicit ones are added.

**Comment by chhsiao90** [comment]
Okay, I will consider if there can be some improvement. Thanks for your feedback.
